rust/actyx/ax-aql/src/language/types.rs

```diff
@@ -8,17 +8,258 @@ use crate::{
 };
 use once_cell::sync::Lazy;
 use pest::pratt_parser::{Assoc, Op, PrattParser};
-use std::sync::Arc;
+use std::{
+    collections::{BTreeMap, BTreeSet},
+    sync::Arc,
+};
 
 #[derive(Debug, PartialEq, Eq, PartialOrd, Ord, Clone)]
 pub enum Type {
+    NoValue,
     Atom(TypeAtom),
     Union(Arc<(Type, Type)>),
     Intersection(Arc<(Type, Type)>),
     Array(Arc<Type>),
     Dict(Arc<Type>),
 }
 
+#[derive(Debug, PartialEq)]
+pub(crate) struct CollapseError(Vec<String>);
+
+impl CollapseError {
+    fn join(errs: Vec<CollapseError>) -> CollapseError {
+        CollapseError(errs.into_iter().flat_map(|x| x.0).collect())
+    }
+}
+
+impl ToString for CollapseError {
+    fn to_string(&self) -> String {
+        self.0.clone().join(". ")
+    }
+}
+
+impl Type {
+    /// Build union out of DoubleEndedIterator of types
+    ///
+    /// # Panics
+    ///
+    /// Panics if the iterator does not yield any value!
+    pub(crate) fn union(iter: impl DoubleEndedIterator<Item = Type>) -> Type {
+        let rebuilt = iter.into_iter().rev().fold(None, |acc, next| match acc {
+            None => Some(next.clone()),
+            Some(prev) => Some(Type::Union(Arc::new((prev, next.clone())))),
+        });
+
+        rebuilt.expect("impossible for union to reduce its subtypes to zero")
+    }
+
+    /// Attempt to apply collapse and sorts to intersections, unions, records
+    pub(crate) fn collapse(self) -> Result<Type, CollapseError> {
```

Comment:
I think this is usually called unification, see: https://go.dev/blog/type-inference#solving-type-equations

Furthermore, the documentation doesn't really explain/define what collapsing is, just that the method collapses things.